src/java.base/share/classes/sun/security/pkcs12/MacData.java

```diff
@@ -70,103 +94,218 @@ class MacData {
+        String digestAlgorithmName = digestAlgorithmId.getName();
+
         // Get the digest.
         this.digest = digestInfo[1].getOctetString();
 
-        // Get the salt.
-        this.macSalt = macData[1].getOctetString();
+        if (digestAlgorithmName.equals("PBMAC1")) {
+            PBMAC1Parameters algParams;
+
+            algParams = new PBMAC1Parameters(digestAlgorithmId
+                    .getEncodedParams());
+
+            this.iterations = algParams.getKdfParams().getIterationCount();
+            this.macSalt = algParams.getKdfParams().getSalt();
+            this.kdfHmac = algParams.getKdfParams().getPrfAlgo();
+            this.keyLength = algParams.getKdfParams().getKeyLength();
 
-        // Iterations is optional. The default value is 1.
-        if (macData.length > 2) {
-            this.iterations = macData[2].getInteger();
+            // Implementations MUST NOT accept params that omit keyLength.
+            if (this.keyLength == -1) {
+                throw new IOException("error: missing keyLength field");
+            }
+            this.hmac = algParams.getHmac();
+            this.macAlgorithm = "pbewith" + this.kdfHmac + "and" + this.hmac;
         } else {
-            this.iterations = 1;
+            this.kdfHmac = null;
+            this.hmac = null;
+            this.keyLength = -1;
+            this.macSalt = macData[1].getOctetString();
+            if (macData.length > 2) {
+                this.iterations = macData[2].getInteger();
+            } else {
+                this.iterations = 1;
+            }
+            // Remove "-" from digest algorithm names
+            this.macAlgorithm = "hmacpbe"
+                    + digestAlgorithmName.replace("-", "");
+        }
+    }
+
+    private static byte[] getMac(String macAlgorithm, char[] password,
+            PBEParameterSpec params, byte[] data,
+            String kdfHmac, String hmac, int keyLength)
+            throws NoSuchAlgorithmException, InvalidKeySpecException,
+            InvalidKeyException, InvalidAlgorithmParameterException {
+        SecretKeyFactory skf;
+        SecretKey pbeKey;
+        Mac m;
+
+        PBEKeySpec keySpec;
+
+        // The Hmac has to be extracted from the algorithm name for
+        // PBMAC1 algorithms. For non-PBMAC1 macAlgorithms, the name
+        // and Hmac are the same.
+        if (macAlgorithm.startsWith("pbewith")) {
+            m = Mac.getInstance(hmac);
+            int len = keyLength == -1 ? m.getMacLength()*8 : keyLength;
+            skf = SecretKeyFactory.getInstance("PBKDF2With" +kdfHmac);
+            keySpec = new PBEKeySpec(password, params.getSalt(),
+                    params.getIterationCount(), len);
+        } else {
+            m = Mac.getInstance(macAlgorithm);
+            skf = SecretKeyFactory.getInstance("PBE");
+            keySpec = new PBEKeySpec(password);
+        }
+        pbeKey = skf.generateSecret(keySpec);
```

Comment:
If the line above fails, there is no chance to clean `keySpec`. Create a big try-finally block.

Reply:
Yes.